samza-core/src/main/java/org/apache/samza/serializers/IntermediateMessageSerde.java

```diff
@@ -67,7 +67,20 @@ public IntermediateMessageSerde(Serde userMessageSerde) {
   public Object fromBytes(byte[] bytes) {
     try {
       final Object object;
-      final MessageType type = MessageType.values()[bytes[0]];
+      final MessageType type;
+      try {
+        type = MessageType.values()[bytes[0]];
+      } catch (ArrayIndexOutOfBoundsException e) {
```

Comment:
Is it possible that this serde gets the old message format but `bytes[0]` is still small enough that there is no `ArrayIndexOutOfBoundsException`? In that case, you wouldn't see your message about the upgrade from `0.13.0`.
I think it was good to have this exception message where you had it before in the outer `catch`, since then you could at least warn users in all cases. You might also want to mention in the exception message that the error could just be due to a user serde error, so that users aren't misled into thinking that their issue is definitely the `0.13.0` upgrade issue.

Reply:
Sorry for the late reply. I moved the exception from the outer catch to here because I think:
1. If this exception is triggered, the failure must be caused by the old format. The exception will tell the user exactly what is the issue.
2. The chance of old message having a first byte of 0 to 2 is relatively small. In that case, users will still see the correct exception if they have multiple containers (with multiple exceptions) or if they look at the source code.
3. New message with user serde error won't have this exception about message type. So users won't be misled by it. And I think this kind of failure is majority.